File: Data Structures & Algorithms/palindrome-partitioning/submission-1.py
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        pal = [[False] * n for _ in range(n)]
        for length in range(1, n + 1):
            for start in range(n - length + 1):
                end = start + length - 1
                pal[start][end] = s[start] == s[end] and (length <= 2 or pal[start + 1][end - 1])

        result = []
        path = []

        def dfs(start):
            if start == n:
                result.append([s[left:right] for left, right in path])
                return
            for end in range(start, n):
                if not pal[start][end]:
                    continue
                path.append((start, end + 1))
                dfs(end + 1)
                path.pop()

        dfs(0)
        return result

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-1.rs
impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        let map = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
        let mut result = Vec::new();
        let mut path = String::new();

        fn dfs(index: usize, digits: &[u8], map: &[&str; 10], path: &mut String, result: &mut Vec<String>) {
            if index == digits.len() {
                result.push(path.clone());
                return;
            }
            let letters = map[(digits[index] - b'0') as usize];
            for letter in letters.bytes() {
                path.push(letter as char);
                dfs(index + 1, digits, map, path, result);
                path.pop();
            }
        }

        if !digits.is_empty() {
            dfs(0, digits.as_bytes(), &map, &mut path, &mut result);
        }
        result
    }
}